src/core/thread/mle_types.cpp

```diff
@@ -64,6 +67,69 @@ DeviceMode::InfoString DeviceMode::ToString(void) const
     return string;
 }
 
+//---------------------------------------------------------------------------------------------------------------------
+// DeviceConfig
+
+#if OPENTHREAD_FTD
+
+DeviceConfig::DeviceConfig(void)
+{
+    Clear();
+
+    mPowerSupply            = OPENTHREAD_CONFIG_DEVICE_POWER_SUPPLY;
+    mLeaderWeightAdjustment = kDefaultAdjustment;
+#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+    mIsBorderRouter = true;
+#endif
+}
+
+uint8_t DeviceConfig::CalculateLeaderWeight(void) const
+{
+    static const int8_t kPowerSupplyIncs[] = {
+        kPowerBatteryInc,          // (0) kPowerSupplyBattery
+        kPowerExtenralInc,         // (1) kPowerSupplyExternal
+        kPowerExtenralStableInc,   // (2) kPowerSupplyExternalStable
+        kPowerExtenralUnstableInc, // (3) kPowerSupplyExternalUnstable
+    };
+
+    static_assert(0 == kPowerSupplyBattery, "kPowerSupplyBattery value is incorrect");
+    static_assert(1 == kPowerSupplyExternal, "kPowerSupplyExternal value is incorrect");
+    static_assert(2 == kPowerSupplyExternalStable, "kPowerSupplyExternalStable value is incorrect");
+    static_assert(3 == kPowerSupplyExternalUnstable, "kPowerSupplyExternalUnstable value is incorrect");
+
+    uint8_t weight = kBaseWeight;
+
+    if (mIsBorderRouter)
+    {
+        weight += (mSupportsCcm ? kCcmBorderRouterInc : kBorderRouterInc);
+    }
+
+    if (mIsUnstable)
```

Comment:
The logic here looks reversed from what the text specifies. It is specified as (pseudocode):

```
// apply powerSupply increment always
...

if ( powerSupply != kPowerSupplyExternalUnstable && powerSupply != kPowerSupplyBattery ) { 
          weight += kIsUnstableInc
}
```

So then a device that is externally powered with UPS-backup, but due to some other bug/issue reboots very often, would have weighting 64 + 4 - 4 = 64.

And a battery-powered device (normally weight 56) that due to some bug/issue reboots very often, would remain at weight 56.   (Current code will cause weight 60 to be used, too high.)

Reply:
Thanks. Updated it in the new push.

I miss-interpreted the text assuming when we have the "is-unstable" then we don't apply condition 5 and 7, but it is other way: If condition 5 and 7 do not apply and we have "is-unstable" then apply the weight. I noticed the situation with battery-powered and "unstable" flag (weight become larger). Was considering to bring it up in the JIRA. Now it make sense :)